File: project/{% if language=='go'%}src{%endif %}/my-contract.go
/*
 * {{spdxAndLicense // SPDX-License-Identifier: Apache-2.0 }}
 */

package main

import (
	"encoding/json"
	"fmt"

	"github.com/hyperledger/fabric-contract-api-go/contractapi"
)

// {{asset }}Contract contract for managing CRUD for {{asset }}
type {{asset }}Contract struct {
	contractapi.Contract
}

// {{asset }}Exists returns true when asset with given ID exists in world state
func (c *{{asset }}Contract) {{asset }}Exists(ctx contractapi.TransactionContextInterface, {{assetCamelCase }}ID string) (bool, error) {
	data, err := ctx.GetStub().GetState({{assetCamelCase }}ID)

	if err != nil {
		return false, err
	}

	return data != nil, nil
}

// Create{{asset }} creates a new instance of {{asset }}
func (c *{{asset }}Contract) Create{{asset }}(ctx contractapi.TransactionContextInterface, {{assetCamelCase }}ID string, value string) error {
	exists, err := c.{{asset }}Exists(ctx, {{assetCamelCase }}ID)
	if err != nil {
		return fmt.Errorf("Could not read from world state. %s", err)
	} else if exists {
		return fmt.Errorf("The asset %s already exists", {{assetCamelCase }}ID)
	}

	{{assetCamelCase }} := new({{asset }})
	{{assetCamelCase }}.Value = value

	bytes, _ := json.Marshal({{assetCamelCase }})

	return ctx.GetStub().PutState({{assetCamelCase }}ID, bytes)
}

// Read{{asset }} retrieves an instance of {{asset }} from the world state
func (c *{{asset }}Contract) Read{{asset }}(ctx contractapi.TransactionContextInterface, {{assetCamelCase }}ID string) (*{{asset }}, error) {
	exists, err := c.{{asset }}Exists(ctx, {{assetCamelCase }}ID)
	if err != nil {
		return nil, fmt.Errorf("Could not read from world state. %s", err)
	} else if !exists {
		return nil, fmt.Errorf("The asset %s does not exist", {{assetCamelCase }}ID)
	}

	bytes, _ := ctx.GetStub().GetState({{assetCamelCase }}ID)

	{{assetCamelCase }} := new({{asset }})

	err = json.Unmarshal(bytes, {{assetCamelCase }})

	if err != nil {
		return nil, fmt.Errorf("Could not unmarshal world state data to type {{asset }}")
	}

	return {{assetCamelCase }}, nil
}

// Update{{asset }} retrieves an instance of {{asset }} from the world state and updates its value
func (c *{{asset }}Contract) Update{{asset }}(ctx contractapi.TransactionContextInterface, {{assetCamelCase }}ID string, newValue string) error {
	exists, err := c.{{asset }}Exists(ctx, {{assetCamelCase }}ID)
	if err != nil {
		return fmt.Errorf("Could not read from world state. %s", err)
	} else if !exists {
		return fmt.Errorf("The asset %s does not exist", {{assetCamelCase }}ID)
	}

	{{assetCamelCase }} := new({{asset }})
	{{assetCamelCase }}.Value = newValue

	bytes, _ := json.Marshal({{assetCamelCase }})

	return ctx.GetStub().PutState({{assetCamelCase }}ID, bytes)
}

// Delete{{asset }} deletes an instance of {{asset }} from the world state
func (c *{{asset }}Contract) Delete{{asset }}(ctx contractapi.TransactionContextInterface, {{assetCamelCase }}ID string) error {
	exists, err := c.{{asset }}Exists(ctx, {{assetCamelCase }}ID)
	if err != nil {
		return fmt.Errorf("Could not read from world state. %s", err)
	} else if !exists {
		return fmt.Errorf("The asset %s does not exist", {{assetCamelCase }}ID)
	}

	return ctx.GetStub().DelState({{assetCamelCase }}ID)
}

File: project/{% if language=='typescript'%}src{%endif %}/src/my-contract.ts
/*
 * {{spdxAndLicense // SPDX-License-Identifier: Apache-2.0 }}
 */

import { Context, Contract, Info, Returns, Transaction } from 'fabric-contract-api';
import { {{asset }} } from './{{assetDashSeparator }}';

@Info({title: '{{asset }}Contract', description: '{{description }}' })
export class {{asset }}Contract extends Contract {

    @Transaction(false)
    @Returns('boolean')
    public async {{assetCamelCase }}Exists(ctx: Context, {{assetCamelCase }}Id: string): Promise<boolean> {
        const data: Uint8Array = await ctx.stub.getState({{assetCamelCase }}Id);
        return (!!data && data.length > 0);
    }

    @Transaction()
    public async create{{asset }}(ctx: Context, {{assetCamelCase }}Id: string, value: string): Promise<void> {
        const exists: boolean = await this.{{assetCamelCase }}Exists(ctx, {{assetCamelCase }}Id);
        if (exists) {
            throw new Error(`The {{assetSpaceSeparator }} ${{{assetCamelCase }}Id} already exists`);
        }
        const {{assetCamelCase }}: {{asset }} = new {{asset }}();
        {{assetCamelCase }}.value = value;
        const buffer: Buffer = Buffer.from(JSON.stringify({{assetCamelCase }}));
        await ctx.stub.putState({{assetCamelCase }}Id, buffer);
    }

    @Transaction(false)
    @Returns('{{asset }}')
    public async read{{asset }}(ctx: Context, {{assetCamelCase }}Id: string): Promise<{{asset }}> {
        const exists: boolean = await this.{{assetCamelCase }}Exists(ctx, {{assetCamelCase }}Id);
        if (!exists) {
            throw new Error(`The {{assetSpaceSeparator }} ${{{assetCamelCase }}Id} does not exist`);
        }
        const data: Uint8Array = await ctx.stub.getState({{assetCamelCase }}Id);
        const {{assetCamelCase }}: {{asset }} = JSON.parse(data.toString()) as {{asset }};
        return {{assetCamelCase }};
    }

    @Transaction()
    public async update{{asset }}(ctx: Context, {{assetCamelCase }}Id: string, newValue: string): Promise<void> {
        const exists: boolean = await this.{{assetCamelCase }}Exists(ctx, {{assetCamelCase }}Id);
        if (!exists) {
            throw new Error(`The {{assetSpaceSeparator }} ${{{assetCamelCase }}Id} does not exist`);
        }
        const {{assetCamelCase }}: {{asset }} = new {{asset }}();
        {{assetCamelCase }}.value = newValue;
        const buffer: Buffer = Buffer.from(JSON.stringify({{assetCamelCase }}));
        await ctx.stub.putState({{assetCamelCase }}Id, buffer);
    }

    @Transaction()
    public async delete{{asset }}(ctx: Context, {{assetCamelCase }}Id: string): Promise<void> {
        const exists: boolean = await this.{{assetCamelCase }}Exists(ctx, {{assetCamelCase }}Id);
        if (!exists) {
            throw new Error(`The {{assetSpaceSeparator }} ${{{assetCamelCase }}Id} does not exist`);
        }
        await ctx.stub.deleteState({{assetCamelCase }}Id);
    }

}
